subprojects/execution/src/main/java/org/gradle/internal/execution/steps/StoreSnapshotsStep.java

```diff
@@ -44,7 +44,7 @@ public CurrentSnapshotResult execute(C context) {
         ImmutableSortedMap<String, CurrentFileCollectionFingerprint> finalOutputs = result.getFinalOutputs();
         context.getBeforeExecutionState().ifPresent(beforeExecutionState -> {
             boolean successful = result.getOutcome().isSuccessful();
-            // Only persist history if there was no failure, or some output files have been changed
+            // Do not store snapshots if there was a failure, but the outputs didn't change
```

Comment:
Still no why. Why do we not store the execution history if there was a failure and the outputs didn't change?

Reply:
I'm not sure what you'd like to see here, care to suggest?

Reply:
```suggestion
            // We do not store the history if there was a failure and the outputs did not change, since then the next execution can be incremental.
            // For example the current execution fails because of a compile failure and for the next execution the source file is fixed, so only the one changed source file needs to be compiled.
```